File: Design-Patterns-Using-JavaScript/Prototype/PrototypePatternShape.js
class Shape {
    constructor() {
        this.color = '';
    }

    toString() {
        throw new Error('toString method must be implemented');
    }

    cloneShape() {
        throw new Error('cloneShape method must be implemented');
    }

    clone() {
        return Object.assign(Object.create(Object.getPrototypeOf(this)), this);
    }
}

class Rectangle extends Shape {
    toString() {
        return 'Rectangle.';
    }

    cloneShape() {
        return this.clone();
    }
}

class Circle extends Shape {
    toString() {
        return 'Circle.';
    }

    cloneShape() {
        return this.clone();
    }
}

class ShapeRegistry {
    static addShape(key, value) {
        if(!ShapeRegistry.shapes)
        ShapeRegistry.shapes = new Map();

        ShapeRegistry.shapes.set(key, value);
    }

    static getShape(key) {
        if (ShapeRegistry.shapes.has(key)) {
            return ShapeRegistry.shapes.get(key).cloneShape();
        }
        return null;
    }

    static load() {
        ShapeRegistry.addShape('circle', new Circle());
        ShapeRegistry.addShape('rectangle', new Rectangle());
    }
}

// Client code
ShapeRegistry.load();
const c = ShapeRegistry.getShape('circle');
const r = ShapeRegistry.getShape('rectangle');
console.log(c.toString(), r.toString());

/*
Circle. Rectangle.
*/

File: Design-Patterns-Using-Ruby/Prototype/PrototypePatternShape.rb
class Shape
    attr_accessor :color

    def initialize
        @color = ''
    end

    def to_s
        raise NotImplementedError, "Subclasses must implement the 'to_s' method"
    end

    def clone_shape
        clone
    end
end

class Rectangle < Shape
    def to_s
        'Rectangle.'
    end
end

class Circle < Shape
    def to_s
        'Circle.'
    end
end

class ShapeRegistry
    @@shapes = {}

    def self.add_shape(key, value)
        @@shapes[key] = value
    end

    def self.get_shape(key)
        return @@shapes[key].clone_shape if @@shapes.key?(key)

        nil
    end

    def self.load
        add_shape('circle', Circle.new)
        add_shape('rectangle', Rectangle.new)
    end
end

# Client code
ShapeRegistry.load
c = ShapeRegistry.get_shape('circle')
r = ShapeRegistry.get_shape('rectangle')
puts c
puts r

=begin 
Circle.
Rectangle. 
=end